File: git/git.go
package git

import (
	"fmt"
	"os"
	"os/exec"
	"strings"

	"github.com/jmelahman/tag/semver"
	log "github.com/sirupsen/logrus"
)

func genTagPattern(prefix, suffix string) string {
	tagPattern := "v[0-9]*.[0-9]*.[0-9]*"
	if suffix != "" {
		tagPattern = fmt.Sprintf("%s-%s*", tagPattern, suffix)
	}
	if prefix != "" {
		tagPattern = fmt.Sprintf("%s/%s", prefix, tagPattern)
	}
	return tagPattern
}

func GetLatestSemverTag(prefix, suffix string) (string, error) {
	tagPattern := genTagPattern(prefix, suffix)
	log.WithFields(log.Fields{
		"pattern": tagPattern,
		"prefix":  prefix,
		"suffix":  suffix,
	}).Debug("GetLatestSemverTag")
	cmd := exec.Command("git", "describe", "--tags", "--abbrev=0", "--match", tagPattern)
	output, err := cmd.Output()
	if err != nil {
		log.Debug("GetLatestSemverTag: no tags found matching pattern, returning v0.0.0")
		if prefix == "" {
			return "v0.0.0", nil
		} else {
			return fmt.Sprintf("%s/v0.0.0", prefix), nil
		}
	}

	matchedTag := strings.TrimSpace(string(output))
	log.WithField("matchedTag", matchedTag).Debug("GetLatestSemverTag: git describe matched tag")

	tagsAt, err := ListTagsAt(matchedTag)
	if err != nil {
		log.WithError(err).WithField("matchedTag", matchedTag).Debug("GetLatestSemverTag: error listing tags at")
		if prefix == "" {
			return "v0.0.0", nil
		} else {
			return fmt.Sprintf("%s/v0.0.0", prefix), nil
		}
	}

	log.WithField("count", len(tagsAt)).WithField("matchedTag", matchedTag).Debug("GetLatestSemverTag: found tags at")

	var largestTag string
	var largestVersion *semver.Version

	for _, tag := range tagsAt {
		if tag == "" {
			continue
		}

		version, err := semver.ParseSemver(tag)
		if err != nil {
			log.WithError(err).WithField("tag", tag).Debug("GetLatestSemverTag: failed to parse tag")
			continue
		}

		// Check prefix: if prefix is empty, version.Prefix should be empty; if prefix is set, version.Prefix should be "prefix/"
		expectedPrefix := ""
		if prefix != "" {
			expectedPrefix = prefix + "/"
		}
		if version.Prefix != expectedPrefix {
			log.WithFields(log.Fields{
				"tag":            tag,
				"expectedPrefix": expectedPrefix,
				"actualPrefix":   version.Prefix,
			}).Debug("GetLatestSemverTag: tag prefix mismatch")
			continue
		}

		if suffix != "" && version.PreRelease != suffix {
			log.WithFields(log.Fields{
				"tag":            tag,
				"expectedSuffix": suffix,
				"actualSuffix":   version.PreRelease,
			}).Debug("GetLatestSemverTag: tag suffix mismatch")
			continue
		}

		if largestVersion == nil {
			largestTag = tag
			largestVersion = version
			log.WithField("tag", tag).Debug("GetLatestSemverTag: initializing largest tag")
			continue
		}

		if semver.CompareSemver(version, largestVersion) {
			log.WithFields(log.Fields{
				"newerTag": tag,
				"olderTag": largestTag,
			}).Debug("GetLatestSemverTag: found newer tag")
			largestTag = tag
			largestVersion = version
		}
	}

	log.WithField("latestTag", largestTag).Debug("GetLatestSemverTag: returning latest tag")

	return largestTag, nil
}

// List all git tags
func ListTags(prefix, suffix string) ([]string, error) {
	tagPattern := genTagPattern(prefix, suffix)
	log.WithFields(log.Fields{
		"pattern": tagPattern,
		"prefix":  prefix,
		"suffix":  suffix,
	}).Debug("ListTags")
	cmd := exec.Command("git", "tag", "-l", tagPattern)
	cmd.Stderr = os.Stderr
	tagsOutput, err := cmd.Output()
	if err != nil {
		log.WithError(err).Debug("ListTags: error running git tag -l")
		return nil, err
	}

	tagList := strings.Split(strings.TrimSpace(string(tagsOutput)), "\n")
	log.WithField("count", len(tagList)).Debug("ListTags: git returned tags (before filtering)")

	// Filter by suffix if specified, since git pattern matching might not be precise enough
	if suffix != "" {
		var filteredList []string
		for _, tag := range tagList {
			if tag == "" {
				continue
			}
			version, err := semver.ParseSemver(tag)
			if err != nil {
				log.WithError(err).WithField("tag", tag).Debug("ListTags: failed to parse tag")
				continue
			}
			if version.PreRelease == suffix {
				filteredList = append(filteredList, tag)
				log.WithField("tag", tag).WithField("suffix", suffix).Debug("ListTags: tag matches suffix")
			} else {
				log.WithFields(log.Fields{
					"tag":            tag,
					"expectedSuffix": suffix,
					"actualSuffix":   version.PreRelease,
				}).Debug("ListTags: tag suffix mismatch")
			}
		}
		log.WithFields(log.Fields{
			"count":  len(filteredList),
			"suffix": suffix,
		}).Debug("ListTags: after filtering, found tags matching suffix")
		return filteredList, nil
	}

	log.WithField("count", len(tagList)).Debug("ListTags: returning tags")

	return tagList, nil
}

func ListTagsAt(ref string) ([]string, error) {
	cmd := exec.Command("git", "tag", "--points-at", ref)
	cmd.Stderr = os.Stderr
	tagsOutput, err := cmd.Output()
	if err != nil {
		return nil, err
	}

	tagList := strings.Split(strings.TrimSpace(string(tagsOutput)), "\n")

	return tagList, nil
}

func TagExists(tag string) (bool, error) {
	tagRef := fmt.Sprintf("refs/tags/%s", tag)
	log.WithField("tag", tag).Debug("TagExists: checking if tag exists")
	cmd := exec.Command("git", "show-ref", "--tags", "--quiet", tagRef)
	if err := cmd.Run(); err != nil {
		if exitError, ok := err.(*exec.ExitError); ok && exitError.ExitCode() == 1 {
			log.WithField("tag", tag).Debug("TagExists: tag does not exist")
			return false, nil
		}
		log.WithError(err).WithField("tag", tag).Debug("TagExists: error checking tag")
		return false, err
	}
	log.WithField("tag", tag).Debug("TagExists: tag exists")
	return true, nil
}

func CreateAndPushTag(tag string, remote string) error {
	log.WithField("tag", tag).Debug("CreateAndPushTag: creating tag")
	cmd := exec.Command("git", "tag", tag)
	cmd.Stderr = os.Stderr
	if err := cmd.Run(); err != nil {
		log.WithError(err).WithField("tag", tag).Debug("CreateAndPushTag: error creating tag")
		return fmt.Errorf("failed to create tag: %w", err)
	}

	log.WithFields(log.Fields{
		"tag":    tag,
		"remote": remote,
	}).Debug("CreateAndPushTag: pushing tag to remote")
	cmd = exec.Command("git", "push", "--quiet", remote, tag)
	cmd.Stderr = os.Stderr
	if err := cmd.Run(); err != nil {
		log.WithError(err).WithFields(log.Fields{
			"tag":    tag,
			"remote": remote,
		}).Debug("CreateAndPushTag: error pushing tag")
		return fmt.Errorf("failed to push tag to %s: %w", remote, err)
	}

	log.WithField("tag", tag).Debug("CreateAndPushTag: successfully created and pushed tag")
	return nil
}

func FetchSemverTags(remote string, prefix, suffix string) error {
	// When suffix is specified, greedily fetch all matching tags (git refspecs don't support
	// wildcards in the middle like v*-suffix*). We'll filter by suffix in the code.
	var tagPattern string
	if prefix != "" {
		tagPattern = fmt.Sprintf("refs/tags/%s/v*:refs/tags/%s/v*", prefix, prefix)
	} else {
		tagPattern = "refs/tags/v*:refs/tags/v*"
	}
	log.WithFields(log.Fields{
		"remote":  remote,
		"refspec": tagPattern,
		"suffix":  suffix,
	}).Debug("FetchSemverTags: fetching from remote (suffix will be filtered later)")
	cmd := exec.Command("git", "fetch", "--quiet", "--prune", remote, tagPattern)
	cmd.Stderr = os.Stderr
	if err := cmd.Run(); err != nil {
		log.WithError(err).Debug("FetchSemverTags: error fetching tags")
		return fmt.Errorf("failed to fetch tags from %s: %w", remote, err)
	}
	log.Debug("FetchSemverTags: successfully fetched tags")
	return nil
}

// GetLatestStableSemverTag returns the latest stable (non-pre-release) semver tag
func GetLatestStableSemverTag(prefix string) (string, error) {
	// Get all tags matching the base pattern (without suffix)
	tagPattern := "v[0-9]*.[0-9]*.[0-9]*"
	if prefix != "" {
		tagPattern = fmt.Sprintf("%s/%s", prefix, tagPattern)
	}
	log.WithFields(log.Fields{
		"pattern": tagPattern,
		"prefix":  prefix,
	}).Debug("GetLatestStableSemverTag")

	cmd := exec.Command("git", "tag", "-l", tagPattern)
	cmd.Stderr = os.Stderr
	output, err := cmd.Output()
	if err != nil {
		log.WithError(err).Debug("GetLatestStableSemverTag: error listing tags")
		if prefix == "" {
			return "v0.0.0", nil
		}
		return fmt.Sprintf("%s/v0.0.0", prefix), nil
	}

	tagList := strings.Split(strings.TrimSpace(string(output)), "\n")
	log.WithField("count", len(tagList)).Debug("GetLatestStableSemverTag: found tags")

	var largestTag string
	var largestVersion *semver.Version

	expectedPrefix := ""
	if prefix != "" {
		expectedPrefix = prefix + "/"
	}

	for _, tag := range tagList {
		if tag == "" {
			continue
		}

		version, err := semver.ParseSemver(tag)
		if err != nil {
			log.WithError(err).WithField("tag", tag).Debug("GetLatestStableSemverTag: failed to parse tag")
			continue
		}

		// Skip tags with different prefix
		if version.Prefix != expectedPrefix {
			continue
		}

		// Skip pre-release versions - we only want stable tags
		if version.PreRelease != "" {
			log.WithField("tag", tag).Debug("GetLatestStableSemverTag: skipping pre-release tag")
			continue
		}

		if largestVersion == nil || semver.CompareSemver(version, largestVersion) {
			largestTag = tag
			largestVersion = version
			log.WithField("tag", tag).Debug("GetLatestStableSemverTag: found newer stable tag")
		}
	}

	if largestTag == "" {
		if prefix == "" {
			return "v0.0.0", nil
		}
		return fmt.Sprintf("%s/v0.0.0", prefix), nil
	}

	log.WithField("latestTag", largestTag).Debug("GetLatestStableSemverTag: returning latest stable tag")
	return largestTag, nil
}

// GetTagAtHEAD returns the semver tag at HEAD, or empty string if none exists
func GetTagAtHEAD(prefix, suffix string) (string, error) {
	tagPattern := genTagPattern(prefix, suffix)
	log.WithFields(log.Fields{
		"pattern": tagPattern,
		"prefix":  prefix,
		"suffix":  suffix,
	}).Debug("GetTagAtHEAD")
	cmd := exec.Command("git", "tag", "--points-at", "HEAD", "--list", tagPattern)
	cmd.Stderr = os.Stderr
	output, err := cmd.Output()
	if err != nil {
		log.WithError(err).Debug("GetTagAtHEAD: error checking tags")
		return "", fmt.Errorf("failed to check tags for HEAD: %w", err)
	}
	tags := strings.TrimSpace(string(output))
	if tags == "" {
		log.Debug("GetTagAtHEAD: no tags found at HEAD")
		return "", nil
	}

	// If multiple tags exist, find the largest one
	tagList := strings.Split(tags, "\n")
	var largestTag string
	var largestVersion *semver.Version

	for _, tag := range tagList {
		if tag == "" {
			continue
		}
		version, err := semver.ParseSemver(tag)
		if err != nil {
			log.WithError(err).WithField("tag", tag).Debug("GetTagAtHEAD: failed to parse tag")
			continue
		}

		// Check suffix match if specified
		if suffix != "" && version.PreRelease != suffix {
			continue
		}

		if largestVersion == nil || semver.CompareSemver(version, largestVersion) {
			largestTag = tag
			largestVersion = version
		}
	}

	log.WithField("tag", largestTag).Debug("GetTagAtHEAD: returning tag at HEAD")
	return largestTag, nil
}

// IsAncestor checks if ancestorRef is an ancestor of descendantRef
func IsAncestor(ancestorRef, descendantRef string) (bool, error) {
	log.WithFields(log.Fields{
		"ancestor":   ancestorRef,
		"descendant": descendantRef,
	}).Debug("IsAncestor: checking ancestry")
	cmd := exec.Command("git", "merge-base", "--is-ancestor", ancestorRef, descendantRef)
	err := cmd.Run()
	if err != nil {
		if exitError, ok := err.(*exec.ExitError); ok && exitError.ExitCode() == 1 {
			log.Debug("IsAncestor: not an ancestor")
			return false, nil
		}
		log.WithError(err).Debug("IsAncestor: error checking ancestry")
		return false, err
	}
	log.Debug("IsAncestor: is an ancestor")
	return true, nil
}

func IsHEADAlreadyTagged(prefix, suffix string) (bool, error) {
	tagPattern := genTagPattern(prefix, suffix)
	log.WithFields(log.Fields{
		"pattern": tagPattern,
		"prefix":  prefix,
		"suffix":  suffix,
	}).Debug("IsHEADAlreadyTagged")
	cmd := exec.Command("git", "tag", "--points-at", "HEAD", "--list", tagPattern)
	cmd.Stderr = os.Stderr
	output, err := cmd.Output()
	if err != nil {
		log.WithError(err).Debug("IsHEADAlreadyTagged: error checking tags")
		return false, fmt.Errorf("failed to check tags for HEAD: %w", err)
	}
	tags := strings.TrimSpace(string(output))
	if tags == "" {
		log.Debug("IsHEADAlreadyTagged: no tags found at HEAD")
		return false, nil
	}
	log.WithField("tags", tags).Debug("IsHEADAlreadyTagged: found tags at HEAD")
	// If suffix is specified, verify that at least one tag matches the suffix
	if suffix != "" {
		tagList := strings.Split(tags, "\n")
		for _, tag := range tagList {
			if tag == "" {
				continue
			}
			version, err := semver.ParseSemver(tag)
			if err != nil {
				log.WithError(err).WithField("tag", tag).Debug("IsHEADAlreadyTagged: failed to parse tag")
				continue
			}
			if version.PreRelease == suffix {
				log.WithFields(log.Fields{
					"tag":    tag,
					"suffix": suffix,
				}).Debug("IsHEADAlreadyTagged: tag matches suffix")
				return true, nil
			}
			log.WithFields(log.Fields{
				"tag":            tag,
				"expectedSuffix": suffix,
				"actualSuffix":   version.PreRelease,
			}).Debug("IsHEADAlreadyTagged: tag suffix mismatch")
		}
		log.WithField("suffix", suffix).Debug("IsHEADAlreadyTagged: no tags at HEAD match suffix")
		return false, nil
	}
	log.Debug("IsHEADAlreadyTagged: HEAD is tagged (suffix not specified)")
	return true, nil
}

File: src/git.rs
use std::process::{Command, Stdio};

use anyhow::{Context, Result};

use crate::semver::{self, Version};

/// Builds the glob `git` uses to narrow tags down to plausible semver ones.
/// The results still need parsing, since globs cannot express semver exactly.
fn tag_pattern(prefix: &str, suffix: &str) -> String {
    let mut pattern = String::from("v[0-9]*.[0-9]*.[0-9]*");
    if !suffix.is_empty() {
        pattern = format!("{pattern}-{suffix}*");
    }
    if !prefix.is_empty() {
        pattern = format!("{prefix}/{pattern}");
    }
    pattern
}

/// The `v0.0.0` tag to fall back on when a repository has no matching tags yet.
fn zero_tag(prefix: &str) -> String {
    if prefix.is_empty() {
        "v0.0.0".to_string()
    } else {
        format!("{prefix}/v0.0.0")
    }
}

fn expected_prefix(prefix: &str) -> String {
    if prefix.is_empty() {
        String::new()
    } else {
        format!("{prefix}/")
    }
}

fn git(args: &[&str]) -> Command {
    let mut cmd = Command::new("git");
    cmd.args(args);
    cmd
}

/// Runs a git command, capturing stdout and letting stderr through to the user.
fn capture(args: &[&str]) -> Result<Option<String>> {
    capture_with(args, Stdio::inherit())
}

/// Like [`capture`], but discards stderr. Used for probes whose failure is an
/// expected, handled outcome rather than something to report to the user.
fn capture_quiet(args: &[&str]) -> Result<Option<String>> {
    capture_with(args, Stdio::null())
}

fn capture_with(args: &[&str], stderr: Stdio) -> Result<Option<String>> {
    let output = git(args)
        .stderr(stderr)
        .output()
        .with_context(|| format!("failed to run `git {}`", args.join(" ")))?;
    if !output.status.success() {
        return Ok(None);
    }
    Ok(Some(
        String::from_utf8_lossy(&output.stdout).trim().to_string(),
    ))
}

/// Runs a git command for its exit status alone, distinguishing "no" (exit 1)
/// from an actual failure the way git's query commands report them.
fn succeeded(args: &[&str]) -> Result<bool> {
    let status = git(args)
        .stderr(Stdio::inherit())
        .status()
        .with_context(|| format!("failed to run `git {}`", args.join(" ")))?;
    match status.code() {
        Some(0) => Ok(true),
        Some(1) => Ok(false),
        _ => anyhow::bail!("`git {}` failed: {status}", args.join(" ")),
    }
}

fn lines(output: &str) -> Vec<String> {
    output
        .lines()
        .map(str::trim)
        .filter(|line| !line.is_empty())
        .map(str::to_string)
        .collect()
}

/// Returns the largest tag in `tags` that satisfies `accept`, if any.
fn largest_tag<F>(tags: &[String], accept: F) -> Option<String>
where
    F: Fn(&Version) -> bool,
{
    let mut largest: Option<(String, Version)> = None;

    for tag in tags {
        let version = match semver::parse(tag) {
            Ok(version) => version,
            Err(err) => {
                log::debug!("largest_tag: failed to parse tag {tag}: {err}");
                continue;
            }
        };
        if !accept(&version) {
            log::debug!("largest_tag: skipping tag {tag}");
            continue;
        }
        if largest.as_ref().is_none_or(|(_, best)| version.gt(best)) {
            log::debug!("largest_tag: {tag} is the largest so far");
            largest = Some((tag.clone(), version));
        }
    }

    largest.map(|(tag, _)| tag)
}

pub fn get_latest_semver_tag(prefix: &str, suffix: &str) -> Result<String> {
    let pattern = tag_pattern(prefix, suffix);
    log::debug!("get_latest_semver_tag: pattern={pattern} prefix={prefix} suffix={suffix}");

    // A repository with no matching tags is an ordinary case, not an error.
    let matched = capture_quiet(&["describe", "--tags", "--abbrev=0", "--match", &pattern])?;
    let Some(matched_tag) = matched.filter(|tag| !tag.is_empty()) else {
        log::debug!("get_latest_semver_tag: no tags found matching pattern, returning v0.0.0");
        return Ok(zero_tag(prefix));
    };
    log::debug!("get_latest_semver_tag: git describe matched tag {matched_tag}");

    // `git describe` reports only one of possibly several tags on that commit,
    // so re-examine every tag there and pick the largest.
    let Some(tags_at) = list_tags_at(&matched_tag)? else {
        log::debug!("get_latest_semver_tag: error listing tags at {matched_tag}");
        return Ok(zero_tag(prefix));
    };
    log::debug!(
        "get_latest_semver_tag: found {} tags at {matched_tag}",
        tags_at.len()
    );

    let want_prefix = expected_prefix(prefix);
    let latest = largest_tag(&tags_at, |version| {
        version.prefix == want_prefix && (suffix.is_empty() || version.pre_release == suffix)
    })
    .unwrap_or_default();

    log::debug!("get_latest_semver_tag: returning latest tag {latest}");
    Ok(latest)
}

/// Returns the latest stable (non-pre-release) semver tag.
pub fn get_latest_stable_semver_tag(prefix: &str) -> Result<String> {
    let pattern = tag_pattern(prefix, "");
    log::debug!("get_latest_stable_semver_tag: pattern={pattern} prefix={prefix}");

    let Some(output) = capture(&["tag", "-l", &pattern])? else {
        log::debug!("get_latest_stable_semver_tag: error listing tags");
        return Ok(zero_tag(prefix));
    };

    let tags = lines(&output);
    log::debug!("get_latest_stable_semver_tag: found {} tags", tags.len());

    let want_prefix = expected_prefix(prefix);
    let latest = largest_tag(&tags, |version| {
        version.prefix == want_prefix && version.pre_release.is_empty()
    })
    .unwrap_or_else(|| zero_tag(prefix));

    log::debug!("get_latest_stable_semver_tag: returning latest stable tag {latest}");
    Ok(latest)
}

pub fn list_tags(prefix: &str, suffix: &str) -> Result<Vec<String>> {
    let pattern = tag_pattern(prefix, suffix);
    log::debug!("list_tags: pattern={pattern} prefix={prefix} suffix={suffix}");

    let output = capture(&["tag", "-l", &pattern])?.context("failed to list tags")?;
    let tags = lines(&output);
    log::debug!(
        "list_tags: git returned {} tags (before filtering)",
        tags.len()
    );

    if suffix.is_empty() {
        return Ok(tags);
    }

    // The glob cannot pin the suffix exactly, so filter it properly here.
    let filtered: Vec<String> = tags
        .into_iter()
        .filter(|tag| match semver::parse(tag) {
            Ok(version) => version.pre_release == suffix,
            Err(err) => {
                log::debug!("list_tags: failed to parse tag {tag}: {err}");
                false
            }
        })
        .collect();
    log::debug!("list_tags: {} tags match suffix {suffix}", filtered.len());

    Ok(filtered)
}

/// Returns the tags pointing at `git_ref`, or `None` if the ref cannot be read.
fn list_tags_at(git_ref: &str) -> Result<Option<Vec<String>>> {
    Ok(capture(&["tag", "--points-at", git_ref])?.map(|output| lines(&output)))
}

/// Returns the semver tags at HEAD matching `prefix`/`suffix`.
fn tags_at_head(prefix: &str, suffix: &str) -> Result<Vec<String>> {
    let pattern = tag_pattern(prefix, suffix);
    log::debug!("tags_at_head: pattern={pattern} prefix={prefix} suffix={suffix}");

    let output = capture(&["tag", "--points-at", "HEAD", "--list", &pattern])?
        .context("failed to check tags for HEAD")?;
    Ok(lines(&output))
}

/// Returns the largest semver tag at HEAD, or `None` if there is none.
pub fn get_tag_at_head(prefix: &str, suffix: &str) -> Result<Option<String>> {
    let tags = tags_at_head(prefix, suffix)?;
    if tags.is_empty() {
        log::debug!("get_tag_at_head: no tags found at HEAD");
        return Ok(None);
    }

    let tag = largest_tag(&tags, |version| {
        suffix.is_empty() || version.pre_release == suffix
    });
    log::debug!("get_tag_at_head: returning tag at HEAD {tag:?}");
    Ok(tag)
}

pub fn is_head_already_tagged(prefix: &str, suffix: &str) -> Result<bool> {
    Ok(get_tag_at_head(prefix, suffix)?.is_some())
}

pub fn tag_exists(tag: &str) -> Result<bool> {
    log::debug!("tag_exists: checking if tag {tag} exists");
    let tag_ref = format!("refs/tags/{tag}");
    let exists = succeeded(&["show-ref", "--tags", "--quiet", &tag_ref])?;
    log::debug!("tag_exists: {tag} exists={exists}");
    Ok(exists)
}

/// Checks whether `ancestor_ref` is an ancestor of `descendant_ref`.
pub fn is_ancestor(ancestor_ref: &str, descendant_ref: &str) -> Result<bool> {
    log::debug!("is_ancestor: checking if {ancestor_ref} is an ancestor of {descendant_ref}");
    succeeded(&["merge-base", "--is-ancestor", ancestor_ref, descendant_ref])
}

pub fn create_and_push_tag(tag: &str, remote: &str) -> Result<()> {
    log::debug!("create_and_push_tag: creating tag {tag}");
    let status = git(&["tag", tag]).stderr(Stdio::inherit()).status()?;
    if !status.success() {
        anyhow::bail!("failed to create tag");
    }

    log::debug!("create_and_push_tag: pushing tag {tag} to {remote}");
    let status = git(&["push", "--quiet", remote, tag])
        .stderr(Stdio::inherit())
        .status()?;
    if !status.success() {
        anyhow::bail!("failed to push tag to {remote}");
    }

    log::debug!("create_and_push_tag: successfully created and pushed tag {tag}");
    Ok(())
}

pub fn fetch_semver_tags(remote: &str, prefix: &str, suffix: &str) -> Result<()> {
    // Refspecs cannot express a wildcard in the middle (`v*-suffix*`), so fetch
    // every version tag greedily and filter by suffix later in the process.
    let refspec = if prefix.is_empty() {
        "refs/tags/v*:refs/tags/v*".to_string()
    } else {
        format!("refs/tags/{prefix}/v*:refs/tags/{prefix}/v*")
    };
    log::debug!(
        "fetch_semver_tags: fetching from {remote} (refspec={refspec}, suffix={suffix} \
         will be filtered later)"
    );

    let status = git(&["fetch", "--quiet", "--prune", remote, &refspec])
        .stderr(Stdio::inherit())
        .status()?;
    if !status.success() {
        anyhow::bail!("failed to fetch tags from {remote}");
    }

    log::debug!("fetch_semver_tags: successfully fetched tags");
    Ok(())
}
